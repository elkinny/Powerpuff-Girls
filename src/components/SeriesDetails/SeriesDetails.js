import React from 'react';
import { useSelector } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faStar, faLanguage, faClock, faFilm,
} from '@fortawesome/free-solid-svg-icons';

import style from './style.module.scss';
import { seriesSelector } from '../../store/selectors';
import EpisodesList from './EpisodesList';
import InfoCard from '../InfoCard';

const SeriesDetails = () => {
  const {
    image, name, genres, summary, language, status, type, rating,
  } = useSelector(seriesSelector);

  return (
    <>
      <InfoCard
        image={image}
        name={name}
        picDetails={genres}
        summary={summary}
        details={[
          <>
            <FontAwesomeIcon icon={faLanguage} />
            {language}
          </>,
          <>
            <FontAwesomeIcon icon={faClock} />
            {status}
          </>,
          <>
            <FontAwesomeIcon icon={faFilm} />
            {type}
          </>,
        ]}
        subTitle={(
          <>
            {rating}
            <FontAwesomeIcon icon={faStar} className={style.star} />
          </>
        )}
      />
      <EpisodesList />
    </>
  );
};

SeriesDetails.getProps = async () => {
  const res = await fetch(`${process.env.API_URL}/shows/${process.env.SERIES_ID}?embed=episodes`);
  const data = await res.json();

  if (!data) {
    return {
      props: { initialReduxState: {} },
    };
  }

  return {
    props: {
      initialReduxState: {
        name: data.name,
        image: data.image.original,
        summary: data.summary,
        genres: data.genres,
        language: data.language,
        status: data.status,
        type: data.type,
        rating: data.rating.average,
        episodes: data._embedded.episodes,
      },
    },
  };
};

export default SeriesDetails;
