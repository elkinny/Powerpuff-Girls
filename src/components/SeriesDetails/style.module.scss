@import '../../styles/constants';

.star {
  color: $colour-yellow !important;
}