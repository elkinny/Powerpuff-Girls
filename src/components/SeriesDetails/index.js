import SeriesDetails from './SeriesDetails';

export default SeriesDetails;
