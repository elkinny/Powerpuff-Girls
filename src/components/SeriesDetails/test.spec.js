import React from 'react';
import { render } from '@testing-library/react';

import SeriesDetails from './SeriesDetails';

jest.mock('../InfoCard', () => () => 'InfoCard');
jest.mock('./EpisodesList', () => () => 'EpisodesList');
jest.mock('react-redux', () => ({
  useSelector: () => ({}),
}));
global.fetch = jest.fn(() => Promise.resolve({
  json: () => Promise.resolve({ }),
}));

describe('SeriesDetails', () => {
  test('renders SeriesDetails component', () => {
    const { container } = render(<SeriesDetails />);

    expect(container).toMatchSnapshot();
  });
});

describe('getProps', () => {
  beforeEach(() => {
    fetch.mockClear();
  });

  test('return formed data', async () => {
    fetch.mockImplementationOnce(() => Promise.resolve({
      json: () => Promise.resolve({
        name: 'name',
        image: {
          original: '/img.png',
        },
        summary: 'summary',
        genres: 'genres',
        language: 'language',
        status: 'status',
        type: 'type',
        rating: { average: 1 },
        _embedded: {
          episodes: [],
        },
      }),
    }));

    const result = await SeriesDetails.getProps();

    expect(result).toMatchSnapshot();
  });

  test('return formed data when data is empty', async () => {
    fetch.mockImplementationOnce(() => Promise.resolve({
      json: () => Promise.resolve(null),
    }));

    const result = await SeriesDetails.getProps();

    expect(result).toMatchSnapshot();
  });
});
