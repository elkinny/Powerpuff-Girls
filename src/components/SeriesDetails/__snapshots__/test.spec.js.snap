// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesDetails renders SeriesDetails component 1`] = `
<div>
  InfoCard
  EpisodesList
</div>
`;

exports[`getProps return formed data 1`] = `
Object {
  "props": Object {
    "initialReduxState": Object {
      "episodes": Array [],
      "genres": "genres",
      "image": "/img.png",
      "language": "language",
      "name": "name",
      "rating": 1,
      "status": "status",
      "summary": "summary",
      "type": "type",
    },
  },
}
`;

exports[`getProps return formed data when data is empty 1`] = `
Object {
  "props": Object {
    "initialReduxState": Object {},
  },
}
`;
